ghcide/src/Development/IDE/Plugin/ChangeTypeAction.hs

```diff
@@ -0,0 +1,129 @@
+                                                 SMethod (..),
+                                                 TextDocumentIdentifier (TextDocumentIdentifier),
+                                                 TextEdit (TextEdit), Uri,
+                                                 WorkspaceEdit (WorkspaceEdit),
+                                                 type (|?) (InR))
+import           Text.Regex.TDFA                ((=~))
+
+descriptor :: PluginId -> PluginDescriptor IdeState
+descriptor plId = (defaultPluginDescriptor plId) { pluginHandlers = mkPluginHandler STextDocumentCodeAction codeActionHandler }
+
+codeActionHandler :: PluginMethodHandler IdeState 'TextDocumentCodeAction
+codeActionHandler ideState plId CodeActionParams {_textDocument = TextDocumentIdentifier uri, _context = CodeActionContext (List diags) _} = response $ do
+      nfp <- getNormalizedFilePath plId (TextDocumentIdentifier uri)
+      decls <- getDecls ideState nfp
+      let actions = generateActions uri decls diags
+      pure $ List actions
+
+getDecls :: MonadIO m => IdeState -> NormalizedFilePath -> ExceptT String m [LHsDecl GhcPs]
+getDecls state = handleMaybeM "Error: Could not get Parsed Module"
+    . liftIO
+    . fmap (fmap (hsmodDecls . unLoc . pm_parsed_source))
+    . runAction "changeSignature.GetParsedModule" state
+    . use GetParsedModule
+
+data ChangeSignature = ChangeSignature { expectedType :: Text
+                                       , actualType   :: Text
+                                       , declName     :: Text
+                                       , declSrcSpan  :: Maybe SrcSpan
+                                       , diagnostic   :: Diagnostic
+                                       , uri          :: Uri
+                                       }
+
+-- Needed to trackdown OccNames in signatures
+type SigName p = (HasOccName (IdP (GhcPass p)))
+
+-- | Generate CodeActions from a list of Diagnostics
+generateActions :: SigName p => Uri -> [LHsDecl (GhcPass p)] -> [Diagnostic] -> [Command |? CodeAction]
+generateActions uri = mapMaybe . generateAction uri
+
+-- | Create a CodeAction from a Diagnostic
+generateAction :: SigName p => Uri -> [LHsDecl (GhcPass p)] -> Diagnostic -> Maybe (Command |? CodeAction)
+generateAction uri decls diag = diagnosticToChangeSig uri decls diag >>= changeSigToCodeAction
+
+-- | Convert a diagnostic into a ChangeSignature and add the proper SrcSpan
+diagnosticToChangeSig :: SigName p => Uri -> [LHsDecl (GhcPass p)] -> Diagnostic -> Maybe ChangeSignature
+diagnosticToChangeSig uri decls diag = addSrcSpan decls <$> matchingDiagnostic uri diag
+
+-- | If a diagnostic has the proper message create a ChangeSignature from it
+matchingDiagnostic :: Uri -> Diagnostic -> Maybe ChangeSignature
+matchingDiagnostic uri diag@Diagnostic{_message} = unwrapMatch $ _message =~ ("Expected type: (.+)\n +Actual type: (.+)\n.*\n +In an equation for ‘(.+)’" :: Text)
```

Comment:
minor nitpick: i would move the message to where bindings under `expectedMsg` or something alike